### PR TITLE
crimson/os/seastore/segment_cleaner: cleanup segment_info_set_t

### PR DESCRIPTION
**Pull request by cyx1231st**
## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

File: src/crimson/os/seastore/segment_cleaner.h [hunk 1]
```diff
@@ -20,299 +20,180 @@
...
 
 private:
-  std::vector<std::optional<segment_manager_info_t>> sm_infos;
   segment_map_t<segment_info_t> segments;
 
-  device_segment_id_t journal_segments = 0;
-  size_t total_bytes = 0;
-  size_t avail_bytes = 0;
-  size_t opened_segments = 0;
-  seastore_off_t segment_size = 0;
```
Comment (athanatos) [inline src/crimson/os/seastore/segment_cleaner.h:288]: Add a comment here:

// See segments_info_t::reset() for initialization

**Review by athanatos — APPROVED** [review]
LGTM other than the nit

@xxhdx1985126 Does this look ok to you?

**Comment by xxhdx1985126** [comment]
I think this pr is based on the assumption that all segment devices have the same configuration, e.g., segment_size, block_size; otherwise LGTM.

**Comment by cyx1231st** [comment]
> I think this pr is based on the assumption that all segment devices have the same configuration, e.g., segment_size, block_size;

I don't see different-configuration is supported right now, that's why I simplified it in the SegmentCleaner. If we want to support heterogenous configurations about segment_size and block_size, we need to introduce the complete feature altogether, and implement proper unit tests to prevent any regression.

**Comment by xxhdx1985126** [comment]
> I don't see different-configuration is supported right now, that's why I simplified it in the SegmentCleaner.

It was supported before, seems to be changed since generic DEVICE pr.

**Comment by cyx1231st** [comment]
> It was supported before and was all segment_manager_info_t about, seems to be changed since the generic DEVICE pr.

If you mean when the PR https://github.com/ceph/ceph/pull/43181 was merged into master. I still think my above comments are true:
* There was no unit test to verify multiple segmented devices with different segment-sizes or block-sizes.
* It may be implemented in SegmentCleaner, but this assumption is not consistent in the entire SeaStore, one example is that the TransactionManager still assumes the block size is the same across all devices: https://github.com/xxhdx1985126/ceph/blob/5ee7916886f70a8b6ce3e68091c13c76f61b1eb6/src/crimson/os/seastore/transaction_manager.h#L495-L497  Another example is that there was no way to configure segment-size and block-size differently.

Sorry I was not able to point it out in that PR. I just realized it afterwards.

**Comment by xxhdx1985126** [comment]
I think the issue here is whether we want to support segment devices with different configurations, I added `segment_manager_info_t` as a preparation for that. If we do want to support this feature, I think we'd better keep `segment_manager_info_t` as it is.

**Comment by athanatos** [comment]
I'm not particularly worried about block_size at this time.  While different devices will have different minimum write sizes (there are zns devices with a 32k min), I haven't heard of any cases where 4k reads aren't supported.  Thus, for the purposes of layout for extents and tree blocks, we probably don't need to worry about different devices having different block sizes.

Different segment_sizes, however, we actually do need to support.  Different zns implementations may have wildly different zone count and sizes which will need to be respected.  I'm ok with putting off support for this for a while, though.

**Comment by cyx1231st** [comment]
My opinion is that partial feature support without tests/validations makes it hard for code maintainence and better quality. It is important to clarify the current invarients in order to introduce further changes that also need to be self-contained.

**Comment by athanatos** [comment]
@cyx1231st Yeah, I agree.  I'm entirely fine with merging this simpler version in order to make progress on gc improvements.

**Review by xxhdx1985126 — APPROVED** [review]

**Pull request merged**